luigi/contrib/s3.py

```diff
@@ -403,145 +352,59 @@ def copy(self, source_path, destination_path, threads=100, start_time=None, end_
 
         # If the file isn't a directory just perform a simple copy
         else:
-            self.__copy_multipart(copy_pool, src_bucket, src_key, dst_bucket, dst_key, part_size, **kwargs)
-            # Close the pool
-            copy_pool.close()
-            copy_pool.join()
-
-    def __copy_multipart(self, pool, src_bucket, src_key, dst_bucket, dst_key, part_size, **kwargs):
-        """
-        Copy a single S3 object to another S3 object, falling back to multipart copy where necessary
-
-        NOTE: This is a private method and should only be called from within the `s3.copy` method
-
-        :param pool: The threadpool to put the s3 copy processes onto
-        :param src_bucket: source bucket name
-        :param src_key: source key name
-        :param dst_bucket: destination bucket name
-        :param dst_key: destination key name
-        :param key_size: size of the key to copy in bytes
-        :param part_size: Part size in bytes. Must be >= 5MB and <= 5 GB.
-        :param kwargs: Keyword arguments are passed to the boto function `initiate_multipart_upload`
-        """
+            copy_source = {
+                'Bucket': src_bucket,
+                'Key': src_key
+            }
+            self.s3.meta.client.copy(
+                copy_source, dst_bucket, dst_key, Config=transfer_config, ExtraArgs=kwargs)
```

Comment:
I've been using this for a couple days and it is taking drastically longer to copy lots of small files.

I'm trying to understand this multi-threaded logic for boto3.

Previously, we were assigning async threads to copy individual files. This new version appears to be assigning possibly multiple threads to copy portions of a single file. Am I reading this correctly?

Reply:
https://github.com/boto/boto3/issues/787

Reply:
I'm not sure if i'm possibly trying to supply too many thread. So i'm decreasing my thread count and will report back with results when i have them.

Reply:
Regardless of thread count, i'm getting 6.67 files copied per second.

I'm skeptical that the current threading implementation provides any benefit for the copying of lots of small files.

Reply:
Re-enabled `ThreadPool` logic for using `apply_async` and actually multithreaded copy works again.

Will submit a PR.